language/resource_types.md

```diff
@@ -0,0 +1,502 @@
+
+### ensureable
+
+Creates an `ensure` property with acceptable values of `present` and `absent`, each of which invoke the provider's `create` and `destroy` methods, respectively:
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  ensurable
+end
+```
+
+Ensurable can also take a block, and in that case, the type should define the allowed set of values. For example, the `package` type defines allowed values as:
+
+```ruby
+Puppet::Type.newtype(:package) do
+  ensurable do
+    newvalue(:present)
+    newvalue(:absent)
+    newvalue(:purged)
+    newvalue(:held)
+    newvalue(:latest)
+    newvalue(/./)
+  end
+end
+```
+
+### newproperty
+
+Defines a new property for a type. Puppet will ensure the resource's current state (as retrieved by the provider) matches the property's desired state (as expressed in the manifest).
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  newproperty(:region) do
+  end
+end
+```
+
+### newparam
+
+Defines a new parameter for the type. Puppet does not enforce state for a parameter. Instead parameters specify additional information about how the provider should ensure the resource and its properties are in the correct state.
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  newparam(:user_data) do
+  end
+end
+```
+
+### newmetaparam
+
+Defines a new metaparameter for a type. Puppet has several built-in metaparameters described below. It is uncommon for types to define new metaparams as these generally require puppet core changes to be useful, e.g. `noop`, `tag`, `schedule`, etc.
+
+### validate
+
+Performs per-resource validation at catalog application time. This method is often used to validate related parameters and properties for a single resource, e.g. mutually exclusive properties or if one property requires another. One concrete example is the `file` resource requires the `source` parameter to be set if the `recurse` parameter is set to `remote`:
+
+```ruby
+Puppet::Type.newtype(:file) do
+  validate do
+    self.fail "You cannot specify a remote recursion without a source" if !self[:source] && self[:recurse] == :remote
+  end
+end
+```
+
+### autorelations
+
+Allows a type to define automatic relationships (`before`, `subscribe`, `require`, `notify`) between each instance of the type and instances of a different type. For example, the `exec` type will autorequire any file resource whose `path` matches the exec  `command`. Note the relationship will only be added to the catalog if puppet is managing both ends of the relationship. That way puppet will automatically create the file containing the command to execute, before trying to executing it, regardless of manifest order.
+
```

Comment:
Also `autorequires` predates `manifest ordering`, back when we had [`unguessable_deterministic_key`](https://github.com/puppetlabs/puppet/blob/029dae917ff9b21fbd6388efab7ce8354d9d3d0e/lib/puppet/transaction.rb#L319), which made it impossible to know what order the agent was going to traverse the graph. But then `autorequires` only works for some resources and some of their relationships, which really made things confusing.

That said, I agree with @adrienthebo, we're just looking to define the current behavior in this doc.